### PR TITLE
Merge in Parameter Sweep Tool Demo

### PR DESCRIPTION
**Pull request by k1nshuk**
## Fixes/Resolves:

N/A

## Summary/Motivation:

Add demo notebook for parameter sweep in tutorials

## Changes proposed in this PR:
- Add demo notebook and corresponding assets.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: tutorials/parameter_sweep_demo.ipynb [hunk 1]
```diff
@@ -0,0 +1,766 @@
...
+   "execution_count": null,
+   "id": "a9de9f80",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "subslide"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "??build_model"
```
Comment (avdudchenko) [inline tutorials/parameter_sweep_demo.ipynb:206]: Can you check on whats going on with these

File: tutorials/parameter_sweep_demo.ipynb [hunk 2]
```diff
@@ -0,0 +1,766 @@
...
+   "execution_count": null,
+   "id": "d106ae65",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "subslide"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "??build_outputs"
```
Comment (avdudchenko) [inline tutorials/parameter_sweep_demo.ipynb:220]: Same here

File: tutorials/parameter_sweep_demo.ipynb [hunk 3]
```diff
@@ -0,0 +1,766 @@
...
+   "execution_count": null,
+   "id": "b07c1657",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "subslide"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "??optimize"
```
Comment (avdudchenko) [inline tutorials/parameter_sweep_demo.ipynb:234]: Same here

File: tutorials/parameter_sweep_demo.ipynb [hunk 4]
```diff
@@ -0,0 +1,766 @@
...
+   "execution_count": null,
+   "id": "015e21aa",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "??build_sweep_params"
```
Comment (avdudchenko) [inline tutorials/parameter_sweep_demo.ipynb:248]: Same here?

File: tutorials/parameter_sweep_demo.ipynb [hunk 5]
```diff
@@ -0,0 +1,766 @@
...
+   "execution_count": null,
+   "id": "c648a402",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "    pprint(results_dict)"
```
Comment (avdudchenko) [inline tutorials/parameter_sweep_demo.ipynb]: Remove the tab here

File: tutorials/parameter_sweep_demo.ipynb [hunk 6]
```diff
@@ -0,0 +1,766 @@
...
+   "execution_count": null,
+   "id": "5410d0c3",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "subslide"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "    pprint(results_array)"
```
Comment (avdudchenko) [inline tutorials/parameter_sweep_demo.ipynb]: Remove the tab here as well

**Review by avdudchenko — APPROVED** [review]
Please double check the small issues in jupyer notebook formatting. Not sure whats going on there. I would also provides direct links in the demo to Loop Tool how to, and parallel manager docs.

File: tutorials/parameter_sweep_demo.ipynb (hunk 1 above)
Comment (k1nshuk) [reply tutorials/parameter_sweep_demo.ipynb:206]: That is just an iPython command to show the function definition

File: tutorials/parameter_sweep_demo.ipynb (hunk 2 above)
Comment (k1nshuk) [reply tutorials/parameter_sweep_demo.ipynb:220]: same here

File: tutorials/parameter_sweep_demo.ipynb (hunk 3 above)
Comment (k1nshuk) [reply tutorials/parameter_sweep_demo.ipynb:234]: same here

File: tutorials/parameter_sweep_demo.ipynb (hunk 4 above)
Comment (k1nshuk) [reply tutorials/parameter_sweep_demo.ipynb:248]: same

**Comment by k1nshuk** [comment]
@avdudchenko I added another cell with links to the documentation in the latest commit. We do not have documentation for the parallel manager. This would need to get fixed in a later PR.

**Review by bknueven — APPROVED** [review]
@lbianchi-lbl does the new notebook get automatically tested as part of being in the `tutorials` directory or do we need to add manual hooks somewhere?

**Comment by bknueven** [comment]
@avdudchenko do you have any issues with merging this?

**Comment by avdudchenko** [comment]
> @avdudchenko do you have any issues with merging this?

Nope should be good to go. (assuming all tests pass)

File: tutorials/parameter_sweep_demo.ipynb [hunk 7]
```diff
@@ -0,0 +1,783 @@
...
+    }
+   },
+   "source": [
+    "## Types of Parameter Sweeps\n",
+    "\n",
+    "* In its current form, a user can run 3 types of parameter sweeps\n",
+    "    - Simple parameter sweep\n",
+    "    - **Recursive parameter sweep**\n",
+    "    - Differential parameter sweep\n",
+    "* Samples for the parameter sweep can be generated from a probability distribution or an *n*-dimensional euclidean space where *n* is the number of parameters in the sweep. Latin hypercube sampling is also enabled."
```
Comment (adam-a-a) [inline tutorials/parameter_sweep_demo.ipynb:423]: I think we should add the simplest (most concise) code block possible to demonstrate exactly how to call/run recursive parameter sweep.

File: tutorials/parameter_sweep_demo.ipynb [hunk 8]
```diff
@@ -0,0 +1,783 @@
...
+   },
+   "source": [
+    "## Types of Parameter Sweeps\n",
+    "\n",
+    "* In its current form, a user can run 3 types of parameter sweeps\n",
+    "    - Simple parameter sweep\n",
+    "    - Recursive parameter sweep\n",
+    "    - **Differential parameter sweep**\n",
+    "* Samples for the parameter sweep can be generated from a probability distribution or an *n*-dimensional euclidean space where *n* is the number of parameters in the sweep. Latin hypercube sampling is also enabled."
+   ]
```
Comment (adam-a-a) [inline tutorials/parameter_sweep_demo.ipynb:517]: I think we should add the simplest (most concise) code block possible to demonstrate exactly how to call/run the differential parameter sweep.

File: tutorials/parameter_sweep_demo.ipynb [hunk 9]
```diff
@@ -0,0 +1,783 @@
...
+   "enable_chalkboard": true,
+   "height": "100%",
+   "scroll": true,
+   "slidenumber": true,
+   "theme": "simple"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 5
+}
```
Comment (adam-a-a) [inline tutorials/parameter_sweep_demo.ipynb:889]: may be an empty cell at the end to delete

**Review by adam-a-a — CHANGES_REQUESTED** [review]
LGTM, except I think we need to add a couple of code examples that the new user could run directly, particularly for recursive and differential sweeps.

**Comment by adam-a-a** [comment]
@k1nshuk I don't mean to be a party pooper. I hadn't had a chance to go through the full demo and figured I'd do so now since it was about to be merged in.

File: tutorials/parameter_sweep_demo.ipynb (hunk 7 above)
Comment (k1nshuk) [reply tutorials/parameter_sweep_demo.ipynb:423]: It should be there now.

File: tutorials/parameter_sweep_demo.ipynb (hunk 8 above)
Comment (k1nshuk) [reply tutorials/parameter_sweep_demo.ipynb:517]: Same here. This was a good exercise. I found more things to fix doing this.

File: tutorials/parameter_sweep_demo.ipynb (hunk 9 above)
Comment (k1nshuk) [reply tutorials/parameter_sweep_demo.ipynb:889]: Yep, removed.

**Comment by k1nshuk** [comment]
@adam-a-a Assuming all the tests pass, this should be good to merge in.

**Review by adam-a-a — APPROVED** [review]

**Pull request merged**